drivers/include/periph/spi.h

```diff
@@ -127,6 +131,19 @@ typedef uint_fast8_t spi_t;
 typedef gpio_t spi_cs_t;
 #endif
 
+/**
+ * @brief   Structure representing the return value of @ref spi_get_clk function
+ *
+ * This structure encapsulates the return code and the SPI clock configuration.
+ */
+#ifndef HAVE_SPI_CLK_T
+typedef struct {
+    int err;        /**< Error code indicating the success or failure of the
+                     *   operation */
+    uword_t clk;    /**< SPI clock configuration */
+} spi_clk_t;
```

Comment:
Combining an error code with the clock configuration in a `struct` looks pretty non-idiomatic C to me. Rather an `int spi_get_clk(spi_clk_t *dest, spi_t bus, uint32_t freq_hz)` would be more idiomatic to me. But that would conflict with my proposed mode of backward compatibility.

But to be honest: After having seen lots of users of peripheral APIs with return code checking and graceful error handling intended to be done by the caller, I lost all hope that driver developers will ever do this correctly. If `spi_get_clk()` would just `assert()` sane settings, we likely get better error handling for less bytes in `.text`.

Also, the only way to incorrectly call `spi_get_clk()` to my understanding is calling it with a non-existing value for `spi_t bus` or with a frequency that is so low that no prescaler can be found. The former clearly is a bug on the caller side, the latter is also difficult to recover from at runtime: If the device attached to a given hardware SPI bus and that bus just cannot generate a frequency compatible with that device, other than falling back to bit-baning SPI in software I have no idea to handle this at runtime. And I'd rather have the application developer than explicitly use the soft SPI than falling back at runtime.

Reply:
> Rather an `int spi_get_clk(spi_clk_t *dest, spi_t bus, uint32_t freq_hz)` would be more idiomatic to me.

:+1:

If I am correct, the only error code used in `spi_clk_t::err` is `-EDOM`? And probably no platform will return 0 as `spi_clk_t::clk` from `spi_get_clk`. So why not use as return value 0 in case of error and the clock different from 0 in case of success?

Reply:
The qn908x platform can return 0:
```c
    /* The value stored in DIV is always (divider - 1), meaning that a value of
     * 0 divides by 1. */
    return (spi_clk_t){ .clk = (uint16_t)(divider - 1) };
```
I guess I should not have changed the description from "Opaque type that contains an SPI clock configuration".

The user should not worry about what it contains but just use it to pass the value obtained from `spi_get_clk` to `spi_acquire` or `spi_get_freq` (which returns an int containing the actual frequency value or -EDOM in the idiomatic way).

I will put a better description in the documentation.

Reply:
> The qn908x platform can return 0:

:thinking: too bad